src/core/client.ts

```diff
@@ -227,50 +227,67 @@ class ApiClient<R extends RawApi> {
-                (resolve, reject) => {
-                    function onStreamError(err: unknown) {
-                        abort();
-                        reject(err);
-                    }
-                    const onHttpError = toHttpError(method, sensLogs, reject);
-                    const config = formDataRequired
-                        ? createFormDataPayload(p, onStreamError)
-                        : createJsonPayload(p);
-                    const opts = {
-                        ...this.options.baseFetchConfig,
-                        signal: abortController.signal,
-                        ...config,
-                    };
-                    fetch(url, opts).then((res) => res.json()).then(resolve)
-                        .catch(onHttpError);
-                },
-            );
-            return res;
         }
+        // Regular request
+        const p = payload ?? {};
+        const controller = new AbortController();
+        const abort = combineAborts(controller, signal);
+        const options = { ...opts.baseFetchConfig, signal: controller.signal };
+        return await new Promise<ApiResponse<ApiCallResult<M, R>>>(
+            (resolve, reject) => {
+                let rejected = false; // prevents repeated rejections
+                /** Error handler for `fetch` */
+                function onHttpError(err: unknown) {
+                    if (rejected) return;
+                    rejected = true;
+                    let msg = `Network request for '${method}' failed!`;
+                    if (isTelegramError(err)) {
+                        msg += ` (${err.status}: ${err.statusText})`;
+                    }
+                    if (opts.sensitiveLogs && err instanceof Error) {
+                        msg += ` ${err.message}`;
+                    }
+                    reject(new HttpError(msg, err));
+                }
+                /** Error handler for timeouts and body stream errors */
+                function onOtherError(err: unknown) {
+                    if (rejected) return;
+                    rejected = true;
+                    reject(err);
+                    abort();
+                }
+                // Create fetch config
+                const config = formDataRequired
+                    ? createFormDataPayload(p, onOtherError)
+                    : createJsonPayload(p);
+                // Use same timeouts for requests as Bot API
+                const timeoutHandle = setTimeout(() => {
+                    const msg =
+                        `Request to '${method}' timed out after 500 seconds`;
+                    onOtherError(new Error(msg));
+                }, 500_000); // 500 seconds
+                // Perform request
+                fetch(url, { ...options, ...config }).then((res) => res.json())
+                    .then(resolve)
+                    .catch(onHttpError)
+                    .finally(() => clearTimeout(timeoutHandle));
```

Comment:
This is roughly what I have in mind from the comment in telegram: https://replit.com/join/mmkxuuwfqm-knightniwrem

Reply:
I see. In other words, we need to check `rejected` before we `resolve` the result of `json`? I don't think we can cancel the parsing somehow.

I think the code already works, but we're calling resolve after reject, which is a bad practice.

Reply:
Hmm.. I'm less concerned about resolving after rejecting, than I am about how the behavior looks like to the developers and users of the bot.

Presumably, if `fetch(...)` completes (e.g. sendMessage), then that means the user would see the message in the chat. However, if timeout intercepts this, it should attempt to abort an already-completed fetch operation, and then reject (i.e. throw the api call). From the developer's perspective, they would assume that the sendMessage failed due to timeout, but in reality, the message is present in the chat.

Reply:
We can clear timeout inside first `then`

Reply:
Now I understand what you mean, good point! I did not think of that. I'll fix it :)